File: PROPMANS.ACCOUNT_MGMT/Customer Accounts/NewAddressForm.vb
Imports ComponentFactory.Krypton.Toolkit
Imports BCL

Imports BCL.CustomerAccounts
Imports PROPMANS.BASE_MOD

Public Class NewAddressForm

#Region "Properties"

    Public _customerId As String

    Public Property CustomerId() As String
        Get
            Return _customerId
        End Get
        Set(ByVal Value As String)
            _customerId = Value
        End Set
    End Property

    Public _contactId As String

    Public Property ContactId() As String
        Get
            Return _contactId
        End Get
        Set(ByVal Value As String)
            _contactId = Value
        End Set
    End Property

    Private _addressLocation As AddressContactLocation

    Public Property AddressLocation() As AddressContactLocation
        Get
            Return _addressLocation
        End Get
        Set(ByVal value As AddressContactLocation)
            _addressLocation = value
        End Set
    End Property
#End Region

#Region "Form & Control Events"

    Private Sub NewAddress_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        UISetup()
    End Sub

    Private Sub btnSave_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnSave.Click

        If txtAddressInformation.Text = String.Empty Then
            MessageBox.Show("Please Input Address", "Value Required", MessageBoxButtons.OK, MessageBoxIcon.Information)
        Else
            If btnSave.Text = "&UPDATE" Then
                UpdateRecord()
            Else
                InsertRecord()
            End If

        End If

    End Sub

#End Region

#Region "Local Procedures"

    Private Sub UISetup()

    End Sub

    Private Sub InsertRecord()

        Dim InsertRecord As New CustomerContact

        With InsertRecord

            .CustomerId = Me.CustomerID
            .ContactLoc = ComboBoxAddressType.SelectedValue
            .Details = txtAddressInformation.Text

            If RbAddressYes.Checked Then
                .Active = Active.Yes
            Else
                .Active = Active.No
            End If

            .IsPreffered = chkMakeDefault.Checked

        End With

        CustomerContact.InsertAddress(InsertRecord)

        Me.DialogResult = Windows.Forms.DialogResult.OK

        Me.Close()

    End Sub

    Private Sub UpdateRecord()

        Dim updateRecord As New CustomerContact

        With updateRecord

            .ContactId = Me.ContactId
            .ContactLoc = ComboBoxAddressType.SelectedValue
            .Details = txtAddressInformation.Text

            .IsPreffered = chkMakeDefault.Checked

            If RbAddressYes.Checked Then
                .Active = Active.Yes
            Else
                .Active = Active.No
                .IsPreffered = False
            End If

            CustomerContact.UpdateAddress(updateRecord)

            If RbAddressYes.Checked Then
                If chkMakeDefault.Checked Then
                    CustomerContact.MakeDefaultAddress(Me.CustomerId, Me.ContactId)
                End If
            End If


        End With

       

        Me.DialogResult = Windows.Forms.DialogResult.OK

        Me.Close()

    End Sub

    Public Sub ViewRecord()

        Common.BindComboBoxtoList(ComboBoxAddressType, GetType(AddressContactLocation))
        ComboBoxAddressType.SelectedValue = AddressLocation

    End Sub

    Private Sub btnCancel_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnCancel.Click

        Me.DialogResult = Windows.Forms.DialogResult.Cancel
        Me.Close()

    End Sub

#End Region

    Private Sub RbAddressNo_CheckedChanged(sender As Object, e As EventArgs) Handles RbAddressNo.CheckedChanged
        chkMakeDefault.Visible = RbAddressYes.Checked
    End Sub
End Class

File: AccountManagement/Forms/NewAddressForm.cs

using System;
using System.Collections.Generic;
using System.Windows.Forms;

using EnumLib;
using PROPMANS.BL.AccountManagement;


namespace PROPMANS.VIEW.AccountManagement
{
    public partial class NewAddressForm : ComponentFactory.Krypton.Toolkit.KryptonForm,
        INewAddress
    {
        NewAddressPresenter presenter;
        #region "Form Instance"
        private static NewAddressForm aForm;
        public static NewAddressForm Instance()
        {
            if (aForm == null)
            {
                aForm = new NewAddressForm();
            }
            return aForm;
        }

        private void Form_Disposed(object sender, System.EventArgs e)
        {
            aForm = null;
        }
        #endregion
        public NewAddressForm()
        {
            InitializeComponent();
            this.Disposed += Form_Disposed;
            presenter = new NewAddressPresenter(this);  
        }

        #region PROPERTIES
        private string _customerID;
        public string CustomerID
        {
            get { return _customerID; }
            set { _customerID = value; }
        }
        #endregion

        #region INTERFACE
        private CustomerContact _customerAddress;
        public CustomerContact CustomerAddress
        {
            get
            {
                if (_customerAddress == null) _customerAddress = new CustomerContact();
               
                _customerAddress.ContactType = ContactType.ADDRESS;
                _customerAddress.ContactLocation = (ContactLocation)cboAddressType.SelectedValue;
                _customerAddress.Details =  txtAddressInformation.Text.Trim();                
                if(RbAddressYes.Checked){
                    _customerAddress.Active = ActiveStatus.Yes;
                }
                else { _customerAddress.Active = ActiveStatus.No; }
                if (chkMakeDefault.Checked) {
                    _customerAddress.Preffered = ActiveStatus.Yes;
                }
                else { _customerAddress.Preffered = ActiveStatus.No; }
                
                return _customerAddress;
            }
            set
            {
                if (value != null)
                {
                    cboAddressType.SelectedValue = (Int32)value.ContactLocation;
                    txtAddressInformation.Text = value.Details;
                    if (value.Active.Equals("Yes"))
                    {
                        RbAddressYes.Checked = true;
                    }
                    else { RbAddressNo.Checked = false; }

                    if (value.Preffered.Equals("Yes"))
                    {
                        chkMakeDefault.Checked = true;
                    }
                    else { chkMakeDefault.Checked = false; }
                    _customerID = value.Customer.CustomerID;
                    _customerAddress = value;
                }                

            }
        }

        public void LoadAddressLocation(List<EnumItem> src)
        {
            cboAddressType.DisplayMember = "Name";
            cboAddressType.ValueMember = "Value";
            cboAddressType.DataSource = src;            
        }

        public void DisplaySaveStatus(bool status)
        {
            if (status)
            {
                MessageBox.Show("New Address Successfully Added!");
                this.DialogResult = System.Windows.Forms.DialogResult.OK;
                this.Close();
            }
            else
            {
                MessageBox.Show("Failed to add new Address!");
            }
        }

        public void DisplayUpdateStatus(bool status)
        {
            if (status)
            {
                MessageBox.Show("Address Successfully Updated!");
                this.DialogResult = System.Windows.Forms.DialogResult.OK;
                this.Close();
            }
            else
            {
                MessageBox.Show("Failed to Update Address!");
            }
        }
        #endregion         
       
        #region EVENTS
        private void btnCancel_Click(object sender, EventArgs e)
        {
            this.DialogResult = System.Windows.Forms.DialogResult.Cancel;
            this.Close();
        }

        private void btnSave_Click(object sender, EventArgs e)
        {
            if (this.btnSave.Text == "&SAVE")
            {
                if (MessageBox.Show("ADD NEW ADDRESS?", "CONFIRM SAVING",
                                MessageBoxButtons.YesNo, MessageBoxIcon.Question, MessageBoxDefaultButton.Button2) == DialogResult.Yes)
                {
                    if (IsInputValid())
                    {
                        presenter.SaveAddress(_customerID);
                    }
                }
            }
            else
            {
                if (MessageBox.Show("UPDATE ADDRESS?", "CONFIRM EDITING",
                                MessageBoxButtons.YesNo, MessageBoxIcon.Question, MessageBoxDefaultButton.Button2) == DialogResult.Yes)
                {
                    if (IsInputValid())
                    {
                        presenter.UpdateContact();
                    }
                }
            }            
            
        }
        #endregion

        #region PRIVATE FUNCTION(S)
        private bool IsInputValid(){
            bool result = true;
            errMsg.Clear();
            errMsg.SetError(txtAddressInformation, "");
            if (String.IsNullOrWhiteSpace(txtAddressInformation.Text))
            {
                errMsg.SetError(txtAddressInformation, "Invalid Address!");
                result = false;
            }

            return result;
        }
        #endregion
    }
}
